lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -422,11 +427,31 @@ def _on_set_password_action(self, event: ActionEvent):
         password = event.params.get("password") or generate_password()
         try:
             self._put_admin_user(password)
-            password = self.secrets.get(Scope.APP, f"{user_name}_password")
+            password = self.secrets.get(Scope.APP, f"{user_name}-password")
             event.set_results({f"{user_name}-password": password})
         except OpenSearchError as e:
             event.fail(f"Failed changing the password: {e}")
 
+    def _on_secret_changed(self, event: ActionEvent):
+        """Refresh secret and re-run corresponding actions if needed."""
+        scope_type, scope_name = event.secret.label.split("-")[0:2]
+        if scope_type == "app" and scope_name == self.app.name:
+            scope = Scope.APP
+        elif scope_type == "unit" and scope_name == self.unit.name:
+            scope = Scope.UNIT
+        else:
+            logging.info(f"Secret {event._label} was not relevant for us.")
+            return
+
+        if scope:
+            logging.debug(f"Secret change for {scope}")
+        else:
+            logging.debug(f"Secret change for {event._label}")
+
+        if any([str(tls_secret_name) in event.secret.label for tls_secret_name in CertType]):
+            for relation in self.opensearch_provider.relations:
+                self.opensearch_provider.update_certs(relation.id, event.chain)
```

Comment:
TBD: What are necessary actions once a/certain secrets change?

Reply:
(NOTE the current code is clearly incorrect, it's just to raise discussions.)

Reply:
Hmm.....
I'm just learning how realtions work so I may be wrong here. But according to what I understand, certs should NOT be handled this way anymore (as soon as my next task is completed ;-) ).

The whole point is this.
We don't wanna have sensitive info -- typically such as certificates -- in plain text in places like relations (peer or external).

Thus, soon in external relations (such as peer relations) we'll have no more than a secret "pointer" (ID or label) to refer to.

This is exactly when the secret-changed event comes to play.

Once any of those "secrets-of-my-interest" would change, I get a notification. INSTEAD of the relation-changed event... as the relation (data) won't change anymore when the secret contetns change. The unique secret reference, held in relation data will be precisely the same as before. However, secret contents have changed and MUST be fetched by interested parties.

Long story short: the code you've referenced from the opensearch_tls module is precisely what should to into the secret-changed event handler.

If my understanding is correct on your response above :-)

Reply:
Confirmed live, solution that's been decided on coming soon.